extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java

```diff
@@ -1174,24 +1187,233 @@ private static void collectDialectConfig(String persistenceUnitName,
+                LOG.warnf( "MariaDB specific settings being ignored because the database is not MySQL." );
+            }
+            else {
+                applyOptionalIntegerSetting(dialectConfig.mysql().bytesPerCharacter(), MYSQL_BYTES_PER_CHARACTER,
+                                            puPropertiesCollector);
+                applyOptionalBooleanSetting(dialectConfig.mysql().noBackslashEscapes(), MYSQL_NO_BACKSLASH_ESCAPES,
+                                            puPropertiesCollector);
+            }
+        }
+
+        if ( dialectConfig.oracle().isAnyPropertySet() ) {
+            if ( databaseKind.isPresent() && databaseKind.get() != SupportedDatabaseKind.ORACLE ) {
+                LOG.warnf( "Oracle specific settings being ignored because the database is not Oracle." );
+            }
+            else {
+                applyOptionalBooleanSetting(dialectConfig.oracle().applicationContinuity(), ORACLE_APPLICATION_CONTINUITY,
+                                            puPropertiesCollector);
+                applyOptionalBooleanSetting(dialectConfig.oracle().autonomous(), ORACLE_AUTONOMOUS_DATABASE,
+                                            puPropertiesCollector);
+                applyOptionalBooleanSetting(dialectConfig.oracle().extended(), ORACLE_EXTENDED_STRING_SIZE,
+                                            puPropertiesCollector);
+            }
+        }
+
+        if ( dialectConfig.mssql().isAnyPropertySet() ) {
+            if ( databaseKind.isPresent() && databaseKind.get() != SupportedDatabaseKind.MSSQL ) {
+                LOG.warnf( "SQL Server specific settings being ignored because the database is not SQL Server." );
+            }
+            else {
+                applyOptionalStringSetting( dialectConfig.mssql().compatibilityLevel(), SQL_SERVER_COMPATIBILITY_LEVEL,
+                                            puPropertiesCollector);
+            }
+        }
+    }
+
+    private static Optional<SupportedDatabaseKind> determineDatabaseKind(Optional<String> dbKind, Optional<String> dialect) {
+        if ( dbKind.isPresent() ) {
+            final String databaseKindName = DatabaseKind.normalize( dbKind.get() );
+            final SupportedDatabaseKind resolved = determineDatabaseKind( databaseKindName );
+            if ( resolved != null ) {
+                return Optional.of( resolved );
+            }
+        }
+
+        if ( dialect.isPresent() ) {
+            String lowercaseDialect = dialect.get().toLowerCase(Locale.ROOT);
+            final SupportedDatabaseKind resolved = determineDatabaseKind( lowercaseDialect );
+            if ( resolved != null ) {
+                return Optional.of( resolved );
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    private static SupportedDatabaseKind determineDatabaseKind(String name) {
+        if ( name.contains( "db2" ) ) {
```

Comment:
`DatabaseKind.isDB2(name)` seems more appropriate? Same for others.

Reply:
No, because this is also called with the Dialect name.

Reply:
I think this will not work well for `db-kind` then, because some database kinds (e.g. `postgresql`) have more than one corresponding string (e.g. `pg`). You'll probably need two different classes.

For what it's worth, we already have code that infers the "db product name" (the Hibernate ORM one) from the explicit dialect here:

https://github.com/quarkusio/quarkus/blob/c09e69a2f7411326108465d2e50a94c489defbad/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java#L1137-L1141

So maybe you should base your code on this `dbProductName` instead?

Reply:
> I think this will not work well for db-kind then, because some database kinds (e.g. postgresql) have more than one corresponding string (e.g. pg).

Look at the callers.  One works with a passed "database kind" and manually does the normalize which handles the case you were talking about (multiple names).  The other works with the Dialect name.

> So maybe you should base your code on this dbProductName instead?

Maybe, I'll take a look.

Reply:
Currently this code handles (1) db-kind and (2) explicit Dialect name.

Are you saying that it is legal to specify a dbProductName but not db-kind?

Reply:
So the "problem" with using dbProductName is that, unless Quarkus does this somewhere, there is no simple mapping of dbProductName to Dialect or SupportedDatabaseKind.

Assuming that it is legal to have just dbProductName without either db-kind or dialect, we'd need a method that handles mapping dbProductName to a SupportedDatabaseKind

Reply:
Ok, I added handling based on database product name as well.

Reply:
Thanks. Then if tests pass, we're all set.

FYI, the only way the database product name can be missing is if someone has no assigned datasource (because it's dynamic, see database multi-tenancy). Even then, users will need to set a dialect explicitly, and as long as they use one of the "standard" dialects, the database product name will be inferred from that dialect.